### PR TITLE
Make ping after discovery concurrent

### PR DESCRIPTION
**Pull request by meyskens**
I would suggest not to merge this yet. this is my attempt to make the ping concurrent. It seems to work however I am not confident enough to say it fully works.

```
HOP:    Address                Loss%  Sent    Last     Avg    Best   Worst                                             Packets
  1:|-- 192.168.0.1             0.0%     5     3.8     3.4     2.5     4.0                                               .....
  2:|-- 94.110.96.2             0.0%     5    13.3    13.3    11.4    14.5                                               .....
  3:|-- 212.65.36.195           0.0%     5    16.0    16.0    15.3    16.5                                               .....
  4:|-- 213.19.195.193          0.0%     5    19.1    21.0    18.7    24.8                                               .....
  5:|-- 4.69.140.34            80.0%     5     0.0    35.8     0.0    35.8                                               .????
  6:|-- 212.3.235.202           0.0%     5    36.2    36.2    35.4    36.9                                               .....
  7:|-- 195.154.1.119          80.0%     5     0.0    36.6     0.0    36.6                                               .????
  8:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
  9:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
 10:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
 11:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
 12:|-- 212.47.231.28           0.0%     5    36.3    36.4    35.3    37.0                                               .....
```
(eg not sure if indeed some routers lose 80% or it could be something catching it incorrect)

### CODE REVIEW

File: imcp/icmp.go [hunk 1]
```diff
@@ -53,3 +53,84 @@ func SendIMCP(localAddr string, dst net.Addr, ttl, pid int, timeout time.Duratio
...
+// SendIMCP sends a IMCP to a given destination
+func SendIMCP(localAddr string, dst net.Addr, ttl, pid int, timeout time.Duration, seq int) (hop ICMPReturn, err error) {
+	hop.Success = false
+	start := time.Now()
+	c, err := icmp.ListenPacket("ip4:icmp", localAddr)
+	if err != nil {
+		return hop, err
+	}
+	defer c.Close()
+	c.IPv4PacketConn().SetTTL(ttl)
```
Comment (golangci) [inline imcp/icmp.go]: Error return value of `(*github.com/tonobo/mtr/vendor/golang.org/x/net/ipv4.genericOpt).SetTTL` is not checked

File: imcp/icmp.go [hunk 2]
```diff
@@ -53,3 +53,84 @@ func SendIMCP(localAddr string, dst net.Addr, ttl, pid int, timeout time.Duratio
...
+	c.SetDeadline(time.Now().Add(timeout))
```
Comment (golangci) [inline imcp/icmp.go]: Error return value of `c.SetDeadline` is not checked

File: imcp/icmp.go [hunk 3]
```diff
@@ -53,3 +53,84 @@ func SendIMCP(localAddr string, dst net.Addr, ttl, pid int, timeout time.Duratio
...
+	return hop, err
+}
+
+// listenForSpecific listens for a reply from a specific destination and returns the body if returned
+func listenForSpecific(deadline time.Time, neededPeer string) (string, error) {
+	conn, err := icmp.ListenPacket("ip4:icmp", "0.0.0.0")
+	if err != nil {
+		return "", err
+	}
+	conn.SetDeadline(deadline)
```
Comment (golangci) [inline imcp/icmp.go]: Error return value of `conn.SetDeadline` is not checked

File: imcp/icmp.go [hunk 4]
```diff
@@ -53,3 +55,87 @@ func SendIMCP(localAddr string, dst net.Addr, ttl, pid int, timeout time.Duratio
...
+// SendIMCP sends a IMCP to a given destination
+func SendIMCP(localAddr string, dst net.Addr, pid int, timeout time.Duration, seq int) (hop ICMPReturn, err error) {
+	hop.Success = false
+	start := time.Now()
+	c, err := icmp.ListenPacket("ip4:icmp", localAddr)
+	if err != nil {
+		return hop, err
+	}
+	defer c.Close()
+	c.SetDeadline(time.Now().Add(timeout))
```
Comment (golangci) [inline imcp/icmp.go]: Error return value of `c.SetDeadline` is not checked

**Comment by meyskens** [comment]
```
HOP:    Address                Loss%  Sent    Last     Avg    Best   Worst                                             Packets
  1:|-- 192.168.0.1             0.0%    15     3.5     3.2     2.0     4.2                                     ...............
  2:|-- 94.110.96.2             0.0%    15    16.5    16.5    13.5    18.5                                     ...............
  3:|-- 212.65.36.195           0.0%    15    15.7    17.8    15.5    23.4                                     ...............
  4:|-- 213.19.195.193          0.0%    15    20.5    20.3    16.3    28.1                                     ...............
  5:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
  6:|-- 212.3.235.202           0.0%    15    35.8    39.5    34.3    67.1                                     ...............
  7:|-- 195.154.1.119           0.0%    15    43.2    37.2    32.3    49.3                                     ...............
  8:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
  9:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
 10:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
 11:|-- ???                   100.0%     1     0.0     0.0     0.0     0.0                                                   ?
 12:|-- 212.47.231.28           0.0%    15    35.0    38.4    35.0    42.7                                     ...............
```

Got better results now!

**Comment by tonobo** [comment]
Are there any thoughts how to handle the packets concurrently? I've tested locally with tracking the outstanding packets in a seperate map, long tiime ago. I'm not sure if the time exceeded packet will contain the echo request at all, if yes it would be pretty easy because you're able to pass the id to it it and the sequence is also available.

**Comment by meyskens** [comment]
What I did is open an icmp listener per goroutine, all packets arrive at all goroutines that are active. 
I did not find an echo in the time exceeded, to my knowledge the only way to link it to the list is the IP of the hop. So what I did is not send a low TTL on the follow-up pings. The only issue here is that if the route changes we can't tell without re-running the discover.

**Comment by meyskens** [comment]
I misread bytes as bits, according to Wikipedia it should return the first 8 bytes so enough for the echo to be returned probably, will take a look at it

**Comment by tonobo** [comment]
After fiddeling some time I mentioned the same, in general raw sockets are shitty at all :smile:

**Comment by meyskens** [comment]
So, i messed a bit with the code and noticed standards exist for nothing... some routers return more data than others, it was my own router that made me see this ;) 

So now I send the timed packets again (some routers reject ping), in the return i parse the echo headers which contain just enough data to get the sequence number I gave it and that works :D

**Comment by meyskens** [comment]
I think this is the best working version of the cuncurrency till now :) Feel free to review it, i am confident it works as expected now but am starting to get tired so can overlook things. I will take a night of sleep and read over my code again tomorrow

**Pull request merged**